### PR TITLE
stdenv: switch default from gcc10 to gcc11

### PR DESCRIPTION
**Pull request by fabianhjr**
###### Description of changes

This will mainly affect i686 and ARM which is still using gcc10;

x86_64 (darwin and linux), RISC-V (linux), and aarch64 (darwin) are
already using gcc11 so this will remove their version override.

aarch64 (linux), vc4, and relibc will remain pinned to previous versions
due to unresolved issues.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vcunat** [comment]
There were regressions why aarch64-linux was kept there.  It doesn't sound like you... confirmed those've been resolved?

File: pkgs/top-level/all-packages.nix
Comment (fabianhjr) [inline pkgs/top-level/all-packages.nix:12503]: Unresolved, aarch64 (linux) is kept pinned on gcc9; will leave as draft for others to test.

**Comment by fabianhjr** [comment]
The platforms that changed version would be ARM and i686 (need testing)

I tested x86_64 linux and had to rebuild a couple hundred or so packages and stdenv.

Would appreciate testing on other platforms like aarch64-linux (gcc9), x86_64-darwin, aarch64-darwin, riscv-linux, etc.

**Comment by fabianhjr** [comment]
> I don't understand this line. Oh you meant you rebuilt those because you based on staging? (as this commit doesn't switch that)

This commit is on top of master (not staging) unsure why I had a stdenv rebuild and a couple hundred rebuilds though.

**Comment by fabianhjr** [comment]
Ah maybe I catched the merge of staging-next into master, its good to know that this doesn't cause a mass-rebuild/stdenv-rebuild.

**Comment by vcunat** [comment]
That had binaries in cache.nixos.org long before the merge.

**Comment by SuperSandro2000** [comment]
If we stay in not red rebuilds we could easily merge this into master if you want. I don't think there is much testing for platforms other the two darwins and linux x86_64 and aarch64 on staging anyway.

**Comment by vcunat** [comment]
`wine` built fine for me, so i686-linux is probably good to go.  I don't know if someone can test 32-bit ARM.

**Review by trofi — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Review by kevincox — APPROVED** [review]

**Comment by misuzu** [comment]
boehm-gc tests are broken on armv7l with gcc11: https://gist.github.com/misuzu/46c4eb4387676124054f7e20a252554b
Maybe related: https://github.com/ivmai/bdwgc/issues/428#issuecomment-1159126937

**Review by misuzu — APPROVED** [review]
Tested building nix and git with this change on top of 74a1793c659d09d7cf738005308b1f86c90cb59b for armv7l-linux, both native & cross.
The issue with boehm-gc is gone now.

**Pull request merged**
